brainiak/eventseg/event.py

```diff
@@ -0,0 +1,356 @@
+
+        # Initialize variables
+        log_scale = np.zeros(t)
+        log_alpha = np.zeros((t, self.n_events + 1))
+        log_beta = np.zeros((t, self.n_events + 1))
+
+        # Forward pass
+        for t in range(t):
+            if t == 0:
+                log_alpha[0, :] = self._log(self.p_start) + logprob[0, :]
+            else:
+                log_alpha[t, :] = self._log(np.exp(log_alpha[t - 1, :])
+                                            .dot(self.P)) + logprob[t, :]
+
+            log_scale[t] = np.logaddexp.reduce(log_alpha[t, :])
+            log_alpha[t] -= log_scale[t]
+
+        # Backward pass
+        log_beta[-1, :] = self._log(self.p_end) - log_scale[-1]
+        for t in reversed(range(t - 1)):
+            obs_weighted = log_beta[t + 1, :] + logprob[t + 1, :]
+            offset = np.max(obs_weighted)
+            log_beta[t, :] = offset + self._log(
+                np.exp(obs_weighted - offset).dot(self.P.T)) - log_scale[t]
+
+        # Combine and normalize
+        log_gamma = log_alpha + log_beta
+        log_gamma -= np.logaddexp.reduce(log_gamma, axis=1, keepdims=True)
+
+        ll = np.sum(log_scale[:(t - 1)]) + np.logaddexp.reduce(
+            log_alpha[-1, :] + log_scale[-1] + self._log(self.p_end), axis=1)
+
+        log_gamma = log_gamma[:, :-1]
+
+        return log_gamma, ll
+
+    def _log(self, x):
+        """Modified version of np.log that manually sets values <=0 to -inf, to
+        avoid runtime warnings
+
+        Parameters
+        ----------
+        x: ndarray of floats
+            Input to the log function
+
+        Returns
+        -------
+        log_ma: ndarray of floats
+            log of x, with x<=0 values replaced with -inf
+        """
+        log_ma = np.ma.log(x).filled(float("-inf"))
+        return log_ma
+
+    def find_events(self, testing_data, var=None, scramble=False):
+        """After fitting an event segmentation using fit(),
+        this function finds the same sequence of event patterns in a new
+        testing dataset.
+
+        Parameters
+        ----------
+        testing_data: timepoint by voxel ndarray
+            fMRI data to segment based on previously-learned event patterns
+
+        var: float or 1D ndarray of length equal to the number of events
+            default: uses variance that maximized training log-likelihood
+            Variance of the event Gaussians. If scalar, all events are
+            assumed to have the same variance.
+
+        scramble: bool : default False
+            If true, the order of the learned events are shuffled before
+            fitting, to give a null distribution
+
+        Returns
+        -------
+        segments : time by event ndarray
+            The resulting soft segmentation. segments[t,e] = probability
+            that timepoint t is in event e
+
+        test_ll : float
+            Log-likelihood of model fit
+        """
+
+        testing_data = testing_data.T
```

Comment:
To avoid modifying data outside the function, perhaps you can replace the testing_data in line 333 with testing_data.T